tests/src/baseservices/TieredCompilation/TieredVtableMethodTests.cs

```diff
@@ -0,0 +1,235 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Reflection;
+using System.Reflection.Emit;
+using System.Runtime.CompilerServices;
+using System.Text;
+using System.Threading;
+
+public static class TieredVtableMethodTests
+{
+    private const int CallCountPerIteration = 8;
+
+    private static StringBuilder s_expectedCallSequence = new StringBuilder();
+    private static StringBuilder s_actualCallSequence = new StringBuilder();
+
+    private static int Main()
+    {
+        const int Pass = 100, Fail = 101;
+
+        var baseObj = new Base();
+        var derivedObj = new Derived();
+        var derivedForDevirtualizationObj = new DerivedForDevirtualization();
+
+        PromoteToTier1(
+            () => CallVirtualMethod(baseObj),
+            () => CallVirtualMethod(derivedObj),
+            () => CallVirtualMethodForDevirtualization(derivedForDevirtualizationObj),
+            () => CallInterfaceVirtualMethodPolymorhpic(baseObj),
+            () => CallInterfaceVirtualMethodPolymorhpic(derivedObj));
+
+        for (int i = 0; i < 4; ++i)
+        {
+            CallVirtualMethod(baseObj, CallCountPerIteration);
+            CallVirtualMethod(derivedObj, CallCountPerIteration);
```

Comment:
I suggest adding a virtual method with a generic method parameter and call it with both valuetype and reference type arguments.